### PR TITLE
Feat: Add train callbacks

### PR DESCRIPTION
**Pull request by mo-sameh**
Re-implementation of the Callback Handler introduced and discussed in MannLabs/alphapeptdeep#154.

### CODE REVIEW

File: peptdeep/model/model_interface.py [hunk 1]
```diff
@@ -392,12 +442,17 @@ def train_with_warmup(self,
...
                 )
-    
+            
             lr_scheduler.step(epoch=epoch, loss=np.mean(batch_cost))
             if verbose: print(
                 f'[Training] Epoch={epoch+1}, lr={lr_scheduler.get_last_lr()[0]}, loss={np.mean(batch_cost)}'
             )
-        
+            continue_training = self.callback_handler.epoch_callback(
+                epoch=epoch, epoch_loss=np.mean(batch_cost)
```
Comment (GeorgWa) [inline peptdeep/model/model_interface.py:451]: question, is there a reason we provide them as `kwargs` ?
This suggests they are optional, altough they aren't

File: peptdeep/model/model_interface.py [hunk 2]
```diff
@@ -447,7 +502,13 @@ def train(self,
                         **kwargs
                     )
                 if verbose: print(f'[Training] Epoch={epoch+1}, Mean Loss={np.mean(batch_cost)}')
-            
+
+                continue_training = self.callback_handler.epoch_callback(
+                    epoch=epoch, epoch_loss=np.mean(batch_cost)
+                    )
+                if not continue_training:
+                    print(f"Training stopped at epoch {epoch}")
```
Comment (GeorgWa) [inline peptdeep/model/model_interface.py:510]: We usually only do logging through the `logging` module.
Should we keep responsibility for logging to the callback implementation?

**Review by GeorgWa — APPROVED** [review]
Very clean implementation with the `CallbackHandler` class 👍🏻

File: peptdeep/model/model_interface.py (hunk 2 above)
Comment (GeorgWa) [reply peptdeep/model/model_interface.py:510]: Okay 👍😊

File: peptdeep/model/model_interface.py (hunk 1 above)
Comment (GeorgWa) [reply peptdeep/model/model_interface.py:451]: That's sensible

**Review by jalew188 — APPROVED** [review]

**Pull request merged**